juju/juju
# [JUJU-2331] Add cli command for update-secret-backend and associated facade apis

We have add, list, remove... this PR adds a command to update secret backends.
The changes include validating that the new config is ok (ping the backend). The PR also updates the add api to use ping.

When updating backend config, values can be reset. The PR adds default value support for the provider config, so that if a value is reset and a default exists, use that.
To disable token rotation, set the new value to 0.
A backend can also be renamed using name=foo as an option.

We also add a validation check to model-config so that setting a model's secret backend checks that the backend exists.

## Checklist


## QA steps

start a vault server
add a secret backend with an invalid endpoint
```
juju add-secret-backend myvault vault --config ~/vaultcfg.yaml
ERROR backend not reachable: Get "http://10.0.0.77:8201/v1/sys/health?drsecondarycode=299&performancestandbycode=299&sealedcode=299&standbycode=299&uninitcode=299": EOF
```
add the backend with valid config

after each update, run `juju secret-backends --format yaml` to check results

rename the backend
`juju update-secret-backend myvault name=myvault2`

add a token rotation
`juju update-secret-backend myvault2 token-rotate=2m`

remove token rotation
`juju update-secret-backend myvault2 token-rotate=0`

update config
`juju update-secret-backend myvault2 namespace=foo`

reset config
`juju update-secret-backend myvault2 --reset namespace`

update config with an invalid endpoint and check error (as for add)

check model-config validation
```juju model-config secret-backend=foo
ERROR secret backend "foo" not found (not found)
```
`juju model-config secret-backend=myvault2`
`juju model-config secret-backend=auto`
